src/Core/Extensions/DiagnosticsExtensions.cs

```diff
@@ -488,7 +488,11 @@ internal static bool IsEffective(this DiagnosticDescriptor descriptor, SyntaxNod
             }
 
             if (compilationOptions.SpecificDiagnosticOptions.TryGetValue(descriptor.Id, out ReportDiagnostic reportDiagnostic))
-                return reportDiagnostic;
+            {
+                return (reportDiagnostic == Microsoft.CodeAnalysis.ReportDiagnostic.Default)
```

Comment:
I just verified that not only this condition can return `Default` but also the one above and below. I would suggest to put these three `if`s to separate local method.